File: cpp/src/posi.cpp
#include "ic_knockoff/posi.hpp"
#include <algorithm>
#include <cmath>

// pi^2 computed at file scope to avoid repeating the expression
namespace {
    // Use 4*atan(1)*4*atan(1) = pi^2 for portability across all C++ standards
    const double kPi2 = std::acos(-1.0) * std::acos(-1.0);
}
#include <limits>
#include <stdexcept>

namespace ic_knockoff {

double alpha_spending_budget(int t, double Q, SpendingSequence sequence, double gamma) {
    if (t < 1)
        throw std::invalid_argument("alpha_spending_budget: t must be >= 1");
    if (Q <= 0.0 || Q >= 1.0)
        throw std::invalid_argument("alpha_spending_budget: Q must be in (0,1)");
    if (sequence == SpendingSequence::Geometric) {
        if (gamma <= 0.0 || gamma >= 1.0)
            throw std::invalid_argument(
                "alpha_spending_budget: gamma must be in (0,1) for Geometric sequence");
        return Q * (1.0 - gamma) * std::pow(gamma, static_cast<double>(t - 1));
    }
    // Riemann Zeta: q_t = Q * 6 / (pi^2 * t^2)
    const double pi2 = kPi2;
    return Q * 6.0 / (pi2 * static_cast<double>(t) * static_cast<double>(t));
}

std::vector<double> alpha_spending_budgets(int max_t, double Q,
                                           SpendingSequence sequence, double gamma) {
    std::vector<double> budgets(static_cast<std::size_t>(max_t));
    for (int t = 1; t <= max_t; ++t)
        budgets[static_cast<std::size_t>(t - 1)] =
            alpha_spending_budget(t, Q, sequence, gamma);
    return budgets;
}

double knockoff_threshold(const std::vector<double>& W,
                          double q_t,
                          const std::set<std::size_t>& active_poly,
                          int offset) {
    // Collect candidate W values (excluding active_poly indices)
    std::vector<double> W_cand;
    W_cand.reserve(W.size());
    for (std::size_t j = 0; j < W.size(); ++j) {
        if (active_poly.find(j) == active_poly.end())
            W_cand.push_back(W[j]);
    }
    if (W_cand.empty())
        return std::numeric_limits<double>::infinity();

    // Collect unique positive |W| values as threshold candidates
    std::vector<double> candidates;
    for (double w : W_cand)
        if (w != 0.0) candidates.push_back(std::abs(w));
    std::sort(candidates.begin(), candidates.end());
    candidates.erase(std::unique(candidates.begin(), candidates.end()),
                     candidates.end());
    if (candidates.empty())
        return std::numeric_limits<double>::infinity();

    for (double tau : candidates) {
        int n_neg = 0, n_pos = 0;
        for (double w : W_cand) {
            if (w <= -tau) ++n_neg;
            if (w >= tau) ++n_pos;
        }
        double ratio = static_cast<double>(offset + n_neg) /
                       std::max(1, n_pos);
        if (ratio <= q_t)
            return tau;
    }
    return std::numeric_limits<double>::infinity();
}

} // namespace ic_knockoff

File: rust/src/posi.rs
//! PoSI α-spending sequences and knockoff+ threshold computation.
//!
//! Implements Section 3 of the IC-Knock-Poly paper.
//!
//! ## Alpha-spending sequences
//!
//! Two sequences satisfying Σ q_t ≤ Q are provided:
//!
//! - **Riemann Zeta** (default):  `q_t = Q · 6 / (π² · t²)`
//! - **Geometric**:               `q_t = Q · (1−γ) · γ^(t−1)`
//!
//! ## Knockoff+ threshold
//!
//! ```text
//! τ_t = min{ τ > 0 :
//!     (1 + |{j ∉ A_poly : W_j ≤ −τ}|)
//!     / max(1, |{j ∉ A_poly : W_j ≥ τ}|)  ≤ q_t }
//! ```

use std::collections::HashSet;

/// Available alpha-spending sequence types.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SpendingSequence {
    /// Riemann Zeta: q_t = Q · 6 / (π² · t²). Default choice.
    RiemannZeta,
    /// Geometric: q_t = Q · (1−γ) · γ^(t−1).
    Geometric,
}

/// Return the FDR budget q_t for iteration t (1-indexed).
///
/// # Parameters
/// - `t`:        Iteration index (≥ 1).
/// - `q`:        Global target FDR level in (0, 1).
/// - `sequence`: Spending sequence type.
/// - `gamma`:    Geometric decay rate (only used for [`SpendingSequence::Geometric`]).
///
/// # Panics
/// Panics if `t < 1`, `q ∉ (0,1)`, or (`sequence == Geometric` and `gamma ∉ (0,1)`).
pub fn alpha_spending_budget(
    t: usize,
    q: f64,
    sequence: SpendingSequence,
    gamma: f64,
) -> f64 {
    assert!(t >= 1, "alpha_spending_budget: t must be >= 1");
    assert!(q > 0.0 && q < 1.0, "alpha_spending_budget: Q must be in (0,1)");
    match sequence {
        SpendingSequence::RiemannZeta => {
            let t_f = t as f64;
            q * 6.0 / (std::f64::consts::PI * std::f64::consts::PI * t_f * t_f)
        }
        SpendingSequence::Geometric => {
            assert!(
                gamma > 0.0 && gamma < 1.0,
                "alpha_spending_budget: gamma must be in (0,1) for Geometric sequence"
            );
            q * (1.0 - gamma) * gamma.powi(t as i32 - 1)
        }
    }
}

/// Compute q_1, ..., q_{max_t} and return them as a `Vec<f64>`.
pub fn alpha_spending_budgets(
    max_t: usize,
    q: f64,
    sequence: SpendingSequence,
    gamma: f64,
) -> Vec<f64> {
    (1..=max_t)
        .map(|t| alpha_spending_budget(t, q, sequence, gamma))
        .collect()
}

/// Compute the knockoff+ threshold τ_t.
///
/// Returns `f64::INFINITY` if no τ satisfies the FDR condition (nothing selected).
///
/// # Parameters
/// - `w`:          W-statistic vector for all candidate features.
/// - `q_t`:        FDR budget for the current iteration.
/// - `active_poly`: Set of indices already in the active polynomial set
///                  (excluded from computation).
/// - `offset`:     1 for knockoff+ (default), 0 for standard knockoff.
pub fn knockoff_threshold(
    w: &[f64],
    q_t: f64,
    active_poly: &HashSet<usize>,
    offset: i64,
) -> f64 {
    // Candidate W values (excluding active set)
    let w_cand: Vec<f64> = w
        .iter()
        .enumerate()
        .filter(|(j, _)| !active_poly.contains(j))
        .map(|(_, &v)| v)
        .collect();

    if w_cand.is_empty() {
        return f64::INFINITY;
    }

    // Unique positive |W| values as threshold candidates, sorted ascending
    let mut candidates: Vec<f64> = w_cand
        .iter()
        .filter(|&&v| v != 0.0)
        .map(|&v| v.abs())
        .collect();
    candidates.sort_by(|a, b| a.partial_cmp(b).unwrap());
    candidates.dedup();

    if candidates.is_empty() {
        return f64::INFINITY;
    }

    for &tau in &candidates {
        let n_neg = w_cand.iter().filter(|&&v| v <= -tau).count() as i64;
        let n_pos = w_cand.iter().filter(|&&v| v >= tau).count() as i64;
        let ratio = (offset + n_neg) as f64 / 1_i64.max(n_pos) as f64;
        if ratio <= q_t {
            return tau;
        }
    }
    f64::INFINITY
}
